### PR TITLE
Check Parameter Consistency with Appropriate Resolved Catalog

### LINKED ISSUE #153: Check Parameter Consistency
**Extended Description**

As a FedRAMP reviewer, in order to ensure certain a CSP's compliance with FedRAMP's flavor of 800-53, I want validations that ensure parameters with FedRAMP assignments are consistent with the requirement.

NB: FedRAMP has not declared constraints for all parameters (and thus are technically not constrained).

Review discussion in #315 for details and examples of potential issues.

**Preconditions**
None.

**Acceptance Criteria**
- [x] All Schematron assertion messages are declarative statements which affirm the positive test outcome.
- [x] All Schematron assertion diagnostic messages are declarative statements which explain the negative test outcome.
- [x] The Schematron code has no assertion failures when validated using `src/validations/styleguides/sch.sch` using the `basic` phase.
- [x] XSpec unit tests for positive and negative Schematron assertion outcomes accompany all Schematron assertions (where feasible).
- [x] A validation which ensures all SSP parameter definitions for which there is a corresponding constraint in the FedRAMP resolved catalog (e.g., `FedRAMP_rev4_MODERATE-baseline-resolved-profile_catalog.xml`) associated with the SSP's FIPS 199 categorization (`<security-sensitivity-level>`) use a value consonant with the constraint.
- [x] Validation error reporting granularity is per-parameter.

**Story Tasks**
- [x] Create one or more Schematron assertions which compare all SSP `<set-parameter>` `<value>` to the value declared in the corresponding (if present) `<control>` `<param>` `<constraint>` value declared in associated resolved catalog based on the SSP FIPS 199 categorization.
- [x] Create test data when necessary.
- [x] Create XSpec unit tests for all Schematron assertions.
- [x] Update Automation Project_Validations and Consistency Mapping Spreadsheet.

**Definition of Done**
 - [x] Acceptance criteria met
 - [x] Unit test coverage of our code > 95%  
 - [x] Automated code quality checks passed 
 - [x] Security reviewed and reported
 - [x] Reviewed against plain language guidelines
 - [x] Code must be self-documenting
 - [x] No local tech debt
 - [x] Load/performance tests passed – needs to be created/automated
 - [x] Documentation updated
 - [x] Architectural Decision Record completed as necessary for significant design choices
 - [ ] PR reviewed & approved
 - [ ] Source code merged

### PR DESCRIPTION
**Pull request by markXLIX**
closes #153

### CODE REVIEW

File: src/validations/rules/ssp.sch [hunk 1]
```diff
@@ -1645,8 +1645,12 @@
...
                 id="uses-correct-param-descriptions"
                 role="error"
-                test="$profile-catalog-control-params/array:controls/o:control/o:param[@id eq $paramID][o:label eq $paramValue or o:constraint/o:description/o:p eq $paramValue]">A
-                FedRAMP SSP must use correct parameter descriptions.</sch:assert>
+                test="
+                    if ($profile-catalog-control-params/array:controls/o:control/o:param[@id eq $paramID]/o:constraint)
+                    then
+                        ($profile-catalog-control-params/array:controls/o:control/o:param[@id eq $paramID][o:constraint/o:description/o:p eq $paramValue])
+                    else
+                        (true())">A FedRAMP SSP must use correct parameter descriptions.</sch:assert>
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1653]: Shouldn't the message (and related identifiers) say something to the effect that a parameter **value** (`set-parameter/value`) **value** matches the related constraint? The assertion messages (assert and diagnostic) refer to **descriptions**, not **values**.

File: src/validations/rules/ssp.sch [hunk 2]
```diff
@@ -1661,6 +1608,46 @@
                 test="current() = $security-objective-levels">[Section B Check 3.10] A FedRAMP SSP must specify an allowed security objective
                 value.</sch:assert>
         </sch:rule>
+        <sch:rule
+            context="o:control-implementation/o:implemented-requirement/o:statement/o:by-component/o:set-parameter"
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1612]: `set-parameter` elements are children of `implemented-requirement` and its descendant `by-component` elements.

Your context incorrectly states they are strictly subordinate to `by-component`.

Correct the context.

The correct context is `o:control-implementation/o:implemented-requirement/descendant-or-self::o:set-parameter`.

At a subsequent time - and a separate issue - `set-parameter` elements subordinate to `control-implementation` will need attention.

File: src/validations/rules/ssp.sch [hunk 3]
```diff
@@ -1661,6 +1608,46 @@
...
+                                select="@id" />
+                            <xsl:copy-of
+                                select="o:param" />
+                        </o:control>
+                    </xsl:for-each>
+                </array:controls>
+            </sch:let>
+            <sch:let
+                name="controlID"
+                value="../../../@control-id" />
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1637]: Should be `ancestor::@control-id`. And this variable is unnecessary.

File: src/validations/rules/ssp.sch [hunk 4]
```diff
@@ -1661,6 +1608,46 @@
...
+            <sch:let
+                name="paramID"
+                value="@param-id" />
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1640]: This variable is unnecessary - should be `current()/@param-id` in the assertion test.

File: src/validations/rules/ssp.sch (hunk 3 above)
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1637]: Should be `ancestor::o:implemented-requirement/@control-id`. And this variable is unnecessary.

File: src/validations/rules/ssp.sch [hunk 5]
```diff
@@ -1661,6 +1608,46 @@
                 test="current() = $security-objective-levels">[Section B Check 3.10] A FedRAMP SSP must specify an allowed security objective
                 value.</sch:assert>
         </sch:rule>
+        <sch:rule
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1611]: Why was this rule placed within the `fips-199` pattern? It has nothing to do with FIPS 199. The correct pattern is `control-implementation`.

File: src/validations/test/ssp.xspec [hunk 6]
```diff
@@ -4364,6 +4364,283 @@
...
+                        uuid="82038bdc-7f3a-4421-922d-b80b96efc446"
+                        xmlns="http://csrc.nist.gov/ns/oscal/1.0">
+                        <metadata>
+                            <title>title</title>
+                            <last-modified>2022-04-04T13:30:41Z</last-modified>
+                            <version>1</version>
+                            <oscal-version>1.0.2</oscal-version>
+                        </metadata>
+                        <import-profile
+                            href="href" />
```
Comment (GaryGapinski) [inline src/validations/test/ssp.xspec:4382]: This is unnecessary.

File: src/validations/test/ssp.xspec [hunk 7]
```diff
@@ -4364,6 +4364,283 @@
...
+                        <system-characteristics>
+                            <system-id>id</system-id>
+                            <system-name>name</system-name>
+                            <description />
```
Comment (GaryGapinski) [inline src/validations/test/ssp.xspec:4386]: This is unnecessary.

File: src/validations/test/ssp.xspec [hunk 8]
```diff
@@ -4364,6 +4364,283 @@
...
+                                type="this-system"
+                                uuid="1d5c6918-b086-4dd2-b225-b73c8e44d292">
+                                <title />
+                                <description>
+                                    <p>This component is the answer to almost everything</p>
+                                </description>
+                                <status
+                                    state="sample" />
+                            </component>
+                        </system-implementation>
```
Comment (GaryGapinski) [inline src/validations/test/ssp.xspec:4446]: This is unnecessary.

File: src/validations/rules/ssp.sch [hunk 9]
```diff
@@ -1608,6 +1608,50 @@
...
+            context="o:control-implementation/o:implemented-requirement/o:statement/o:by-component/o:set-parameter"
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1612]: The context is too specific. `set-parameter` can occur as a child of `implemented-requirement` and `by-component` elements. It can also be found as a child of `control-implementation` (but consider that in a separate issue).

A context of `o:control-implementation/o:implemented-requirement/descendant-or-self::o:set-parameter` should suffice.

File: src/validations/test/ssp.xspec [hunk 10]
```diff
@@ -4364,6 +4364,283 @@
...
+                                        <set-parameter param-id="ma-1_prm_2">
+                                            <value>at least every 3 years</value>
+                                        </set-parameter>
+                                        <set-parameter param-id="ma-1_prm_3">
+                                            <value>at least annually</value>
+                                        </set-parameter>
+                                    </by-component>
+                                </statement>
+                            </implemented-requirement>
+                        </control-implementation>
```
Comment (GaryGapinski) [inline src/validations/test/ssp.xspec:4498]: https://github.com/18F/fedramp-automation/pull/436/commits/7c117d479698f437841b29736e9219ac98520404 was created in order to remove `set-parameter` elements from `statement` elements.

These `set-parameter` elements should be child elements of `implemented-requirement` **as well as** subordinate to `by-component`. The unit test should ensure both variants are correctly handled.

File: src/validations/test/ssp.xspec [hunk 11]
```diff
@@ -4364,6 +4364,283 @@
...
+                <x:context>
+                    <system-security-plan
+                        uuid="82038bdc-7f3a-4421-922d-b80b96efc446"
+                        xmlns="http://csrc.nist.gov/ns/oscal/1.0">
+                        <metadata>
+                            <title>title</title>
+                            <last-modified>2022-04-04T13:30:41Z</last-modified>
+                            <version>1</version>
+                            <oscal-version>1.0.2</oscal-version>
+                        </metadata>
```
Comment (GaryGapinski) [inline src/validations/test/ssp.xspec:4516]: This is superfluous.

File: src/validations/test/ssp.xspec [hunk 12]
```diff
@@ -4364,6 +4364,283 @@
...
+                        <import-profile
+                            href="href" />
+                        <system-characteristics>
+                            <system-id>id</system-id>
+                            <system-name>name</system-name>
+                            <description />
```
Comment (GaryGapinski) [inline src/validations/test/ssp.xspec:4522]: This is superfluous.

File: src/validations/test/ssp.xspec [hunk 13]
```diff
@@ -4364,6 +4364,283 @@
...
+                                <security-objective-integrity>fips-199-moderate</security-objective-integrity>
+                                <security-objective-availability>fips-199-moderate</security-objective-availability>
+                            </security-impact-level>
+                            <status
+                                state="here" />
+                            <authorization-boundary>
+                                <description />
+                                <diagram
+                                    uuid="82038bdc-7f3a-4421-922d-b80b96efc446" />
+                            </authorization-boundary>
```
Comment (GaryGapinski) [inline src/validations/test/ssp.xspec:4567]: This is superfluous.

File: src/validations/test/ssp.xspec [hunk 14]
```diff
@@ -4364,6 +4364,283 @@
...
+                                type="this-system"
+                                uuid="1d5c6918-b086-4dd2-b225-b73c8e44d292">
+                                <title />
+                                <description>
+                                    <p>This component is the answer to almost everything</p>
+                                </description>
+                                <status
+                                    state="sample" />
+                            </component>
+                        </system-implementation>
```
Comment (GaryGapinski) [inline src/validations/test/ssp.xspec:4582]: This is superfluous.

File: src/validations/test/ssp.xspec [hunk 15]
```diff
@@ -4364,6 +4364,283 @@
...
+                                        <set-parameter param-id="ma-1_prm_2">
+                                            <value>[defined the length of time]</value>
+                                        </set-parameter>
+                                        <set-parameter param-id="ma-1_prm_3">
+                                            <value>at least annually</value>
+                                        </set-parameter>
+                                    </by-component>
+                                </statement>
+                            </implemented-requirement>
+                        </control-implementation>
```
Comment (GaryGapinski) [inline src/validations/test/ssp.xspec:4634]: This tests `set-parameter` within `implemented-requirement` but not when `set-parameter` is a child of `implemented-requirement`.

File: src/validations/test/ssp.xspec [hunk 16]
```diff
@@ -4372,6 +4649,57 @@
...
+                                    </adjustment-justification>
+                                </availability-impact>
+                            </information-type>
+                        </system-information>
+                        <!-- Table 2-3 Security Impact Level -->
+                        <security-impact-level>
+                            <security-objective-confidentiality>fips-199-moderate</security-objective-confidentiality>
+                            <security-objective-integrity>fips-199-moderate</security-objective-integrity>
+                            <security-objective-availability>fips-199-moderate</security-objective-availability>
+                        </security-impact-level>
```
Comment (GaryGapinski) [inline src/validations/test/ssp.xspec:4702]: Why is this here?

File: src/validations/rules/ssp.sch [hunk 17]
```diff
@@ -3866,16 +3866,6 @@
...
-        <sch:diagnostic
-            doc:assertion="contingency-planning-for-information-system-backup"
-            doc:context="/o:system-security-plan/o:control-implementation/o:implemented-requirement"
-            id="contingency-planning-for-information-system-backup-diagnostic">The statement <xsl:value-of
-                select="@statement-id" /> does not contain correct interval terms for system backups.</sch:diagnostic>
-        <sch:diagnostic
-            doc:assertion="contingency-planning-for-information-system-backup2"
-            doc:context="/o:system-security-plan/o:control-implementation/o:implemented-requirement"
-            id="contingency-planning-for-information-system-backup2-diagnostic">The statement <xsl:value-of
-                select="@statement-id" /> does not contain correct scope terms for system backups.</sch:diagnostic>
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch]: Why was this not removed when the related assertions were removed?

File: src/validations/rules/ssp.sch (hunk 1 above)
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1653]: The issue which spawned your PR says the following but you have not accomplished that task.

> Create one or more Schematron assertions which compare all SSP `<implemented-requirement>` `<set-parameter>` `<value>` to the value declared in the corresponding (if present) `<control>` `<param>` `<constraint>` value declared in associated resolved catalog based on the SSP FIPS 199 categorization.

File: src/validations/rules/ssp.sch [hunk 18]
```diff
@@ -1661,6 +1608,46 @@
...
+                        <o:control>
+                            <xsl:attribute
+                                name="id"
+                                select="@id" />
+                            <xsl:copy-of
+                                select="o:param" />
+                        </o:control>
+                    </xsl:for-each>
+                </array:controls>
+            </sch:let>
```
Comment (GaryGapinski) [inline src/validations/rules/ssp.sch:1634]: This variable is unnecessary.

File: src/validations/rules/ssp.sch (hunk 17 above)
Comment (markXLIX) [reply src/validations/rules/ssp.sch]: This section of code is not in the ssp.sch.

**Review by GaryGapinski — CHANGES_REQUESTED** [review]
Start over. This PR is littered with superfluous commits, errors, and poor coding practices.

- Read the original issue #153. Pay attention to the story tasks.
- Refer to the OSCAL documentation regarding `set-parameter` for the locations at which it can be employed. Use that information when coding.
- Do not employ unnecessary variables.
- Find the correct Schematron pattern and use it.
- Read the assertion and diagnostic messages you create to ensure they are appropriate.
- Use appropriately-named identifiers (i.e., ones which correspond to the intent of the assertion(s)).
- Test the Schematron using the `advanced` phase of `sch.sch`.
- Read the original issue #153. Pay attention to the acceptance criteria. Check the boxes corresponding to the acceptance criteria, tasks, and definition of done items which you have accomplished at the time you intend to submit a pull request but prior to submission.
